rcl/src/rcl/rmw_implementation_identifier_check.c

```diff
@@ -63,16 +89,54 @@ INITIALIZER(initialize) {
     )
     exit(ret);
   }
-  // If the environment variable is set, and it does not match, print a warning and exit.
-  if (strlen(expected) > 0 && strcmp(rmw_get_implementation_identifier(), expected) != 0) {
+  if (strlen(asserted_rmw_impl_env) > 0) {
+    // Copy the environment variable so it doesn't get over-written by the next getenv call.
+    asserted_rmw_impl = (char *)allocator.allocate(
+      strlen(asserted_rmw_impl_env) + 1, allocator.state);
+    if (!asserted_rmw_impl) {
+      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "allocation failed")
+      exit(RCL_RET_BAD_ALLOC);
+    }
+    memcpy(asserted_rmw_impl, asserted_rmw_impl_env, strlen(asserted_rmw_impl_env) + 1);
+  }
+
+  // If both environment variables are set, and they do not match, print a warning and exit.
+  if (expected_rmw_impl && asserted_rmw_impl && strcmp(expected_rmw_impl, asserted_rmw_impl) != 0) {
     RCUTILS_LOG_ERROR_NAMED(
       ROS_PACKAGE_NAME,
-      "Expected RMW implementation identifier of '%s' but instead found '%s', exiting with %d.",
-      expected,
-      rmw_get_implementation_identifier(),
-      RCL_RET_MISMATCHED_RMW_ID
+      "Values of RMW_IMPLEMENTATION ('%s') and RCL_ASSERT_RMW_ID_MATCHES ('%s') environment "
+      "variables do not match, exiting with %d.",
+      expected_rmw_impl, asserted_rmw_impl, RCL_RET_ERROR
     )
-    exit(RCL_RET_MISMATCHED_RMW_ID);
+    exit(RCL_RET_ERROR);
+  }
+
+  // Collapse the expected_rmw_impl and asserted_rmw_impl variables so only expected_rmw_impl needs
+  // to be used from now on.
+  if (expected_rmw_impl && asserted_rmw_impl) {
+    // The strings at this point must be equal.
+    // No need for asserted_rmw_impl anymore, free the memory.
+    allocator.deallocate((char *)asserted_rmw_impl, allocator.state);
+  } else {
+    if (asserted_rmw_impl) {
+      expected_rmw_impl = asserted_rmw_impl;
+    }
+  }
+
+  // If either environment variable is set, and it does not match, print a warning and exit.
+  if (expected_rmw_impl) {
+    if(strcmp(rmw_get_implementation_identifier(), expected_rmw_impl) != 0) {
```

Comment:
Can `rmw_get_implementation_identifier()` return null here? Maybe not, but perhaps it would be best to be defensive and check it.

Reply:
circling back, it can return null if an invalid rmw implemenation is requested: https://github.com/ros2/rcl/pull/201#issuecomment-349770216 shows it used to segfault if requesting and asserting an invalid rmw implementation, so it's a good thing we checked this; thanks for the suggestion